### PR TITLE
[CELEBORN-1444] Introduce worker decommission metrics and corresponding REST API

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?

Introduce worker decommission metrics and corresponding REST API.

### Why are the changes needed?

In a production environment, due to certain hardware or environmental reasons, our script will automatically decommission the node. At this time, we need to distinguish between graceful shutdown nodes and decommissioned nodes.

If we distinguish shutdown worker and decommission worker metrics, we can achieve better operation and maintenance.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

- `DefaultMetaSystemSuiteJ#testHandleReportWorkerDecommission`
- `RatisMasterStatusSystemSuiteJ#testHandleReportWorkerDecommission`
- `ApiMasterResourceSuite#decommissionWorkers`
- `ApiWorkerResourceSuite#isDecommissioning`

### CODE REVIEW

**Comment by leixm** [comment]
@SteNicholas  @AngersZhuuuu  Can you help review?

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 1]
```diff
@@ -436,6 +446,12 @@ public void updateWorkerEventMeta(int workerEventTypeValue, List<WorkerInfo> wor
     }
   }
 
+  public void updateMetaByReportWorkerDecommission(List<WorkerInfo> workers) {
+    synchronized (this.workers) {
+      decommissionWorkers.addAll(workers);
```
Comment (RexXiong) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:451]: Before this, worker decommission will ReportWorkerUnavailable, which indicates the worker would be shutdown, then Client would quickly tell worker 
commits those associated partitions,  If we change shutdownWorkers to decommissionWorkers, May be decommission take longer time than before.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 2]
```diff
@@ -971,6 +986,14 @@ private[celeborn] class Worker(
     }
     serverBootstraps
   }
+
+  private def isDecommissioning: Int = {
+    if (shutdown.get() && workerStatusManager.exitEventType == WorkerEventType.Decommission) {
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:991]: Wd need use currentWorkerStatus.getState to check the worker status.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java (hunk 1 above)
Comment (leixm) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:451]: It seems that we should return decommissionWorkers to the client through heartbeat, but do we need to distinguish between shutdown and decommission on the client side?

Comment (RexXiong) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:451]: > It seems that we should return decommissionWorkers to the client through heartbeat, but do we need to distinguish between shutdown and decommission on the client side?

Currently, the client doesn't need to know that.

Comment (AngersZhuuuu) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:451]: Just. return decommission worker + shutdown worker to client side?

Comment (leixm) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:451]: > Just. return decommission worker + shutdown worker to client side?

I think so too

File: common/src/main/proto/TransportMessages.proto [hunk 3]
```diff
@@ -789,4 +791,9 @@ message PbPackedPartitionLocationsPair {
...
   PbPackedPartitionLocationsPair locationPairs = 1;
   string networkLocation = 2;
-}
\ No newline at end of file
+}
+
+message PbReportWorkerDecommission {
+  repeated PbWorkerInfo workers = 1;
+  string requestId = 2;
+}
```
[the remark this answers is not in the document]
Comment (leixm) [reply common/src/main/proto/TransportMessages.proto:799]: Blane line already exists, it seems to be a diff display problem.

File: METRICS.md [hunk 4]
```diff
@@ -147,6 +148,7 @@ Here is an example of Grafana dashboard importing.
 |         PotentialConsumeSpeed          |      worker       |                        This value means speed of potential consumption for congestion control.                         |
 |            UserProduceSpeed            |      worker       |                           This value means speed of user production for congestion control.                            |
 |           WorkerConsumeSpeed           |      worker       |                          This value means speed of worker consumption for congestion control.                          |
+|        isDecommissioningWorker         |      worker       |                              Show if the worker is during the process of decommissioning.                              |
```
[the remark this answers is not in the document]
Comment (leixm) [reply METRICS.md:151]: Sometimes we want to pay attention to which node decommission has not been completed yet.

[the remark this answers is not in the document]
Comment (leixm) [reply METRICS.md:151]: isExiting is also good. Maybe we can distinguish between graceful shutdown and decommission.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HAMasterMetaManager.java [hunk 5]
```diff
@@ -365,6 +365,27 @@ public void handleWorkerEvent(
...
+              .setReportWorkerDecommissionRequest(
+                  ResourceProtos.ReportWorkerDecommissionRequest.newBuilder()
+                      .addAllWorkers(addrs)
+                      .build())
+              .build());
+    } catch (CelebornRuntimeException e) {
+      LOG.error("Handle worker decommission failed!", e);
+      throw e;
+    }
+    updateMetaByReportWorkerDecommission(workers);
```
Comment (RexXiong) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HAMasterMetaManager.java:386]: No need extra call updateMetaByReportWorkerDecommission

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 6]
```diff
@@ -971,6 +986,14 @@ private[celeborn] class Worker(
     }
     serverBootstraps
   }
+
+  private def isDecommissioning: Int = {
+    if (shutdown.get() && workerStatusManager.currentWorkerStatus.getState == State.InDecommission) {
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:991]: State `InDecommissionThenIdle` should also be considered as `isDecommissioning`.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java (hunk 1 above)
Comment (RexXiong) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:451]: Looks good!

**Comment by leixm** [comment]
Seems flaky test.

File: METRICS.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply METRICS.md:151]: In the doc, need desc each value's meanning, such as  1 means is during decommission

Comment (AngersZhuuuu) [inline METRICS.md:151]: We can also add a restful API for listing workers during decommission

**Comment by leixm** [comment]
@AngersZhuuuu  @RexXiong  @SteNicholas  Can you help review?

File: docs/monitoring.md [hunk 7]
```diff
@@ -369,6 +371,7 @@ API path listed as below:
...
 | /shutdownWorkers     | GET    |                                              | List all shutdown workers of the master.                                                                                                                                                                                                                                                                                                                     |
+| /decommissionWorkers | GET    |                                              | List all decommission workers of the master.                                                                                                                                                                                                                                                                                                                 |
```
Comment (AngersZhuuuu) [inline docs/monitoring.md:374]: Also add `isDecommission` restful API for worker.

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/http/api/ApiWorkerResource.scala [hunk 8]
```diff
@@ -56,6 +56,15 @@ class ApiWorkerResource extends ApiRequestContext {
   @GET
   def isShutdown: String = httpService.isShutdown
 
+  @Path("/isDecommissioning")
```
Comment (SteNicholas) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/http/api/ApiWorkerResource.scala:59]: Add the test case for this REST API.

**Comment by SteNicholas** [comment]
@leixm, please update the title and description of this pull request including the introduced REST API.

File: assets/grafana/celeborn-dashboard.json [hunk 9]
```diff
@@ -1357,34 +1357,19 @@
           ],
           "title": "metrics_ShutdownWorkerCount_Value",
           "type": "timeseries"
-        }
```
Comment (SteNicholas) [inline assets/grafana/celeborn-dashboard.json:1360]: Have you ever verified the dashboard?

Comment (leixm) [reply assets/grafana/celeborn-dashboard.json:1360]: SeaTalk_IMG_20240606_111048
SeaTalk_IMG_20240606_111302

Comment (leixm) [reply assets/grafana/celeborn-dashboard.json:1360]: The monitoring of the panel is normal, but I found that there is a problem with the layout. This problem also existed before this PR. I think we may need to re-export the entire dashboard. For example, when I click the overall title, part of the overall panel will pop up. Go to the top of the dashboard.

Comment (leixm) [reply assets/grafana/celeborn-dashboard.json:1360]: When i import
WechatIMG84
Click overall
WechatIMG85
Click Master
SeaTalk_IMG_20240606_112033

**Comment by leixm** [comment]
@SteNicholas  Can you help review?

**Review by RexXiong — COMMENTED** [review]
Overall LGTM, need add UT in 
DefaultMetaSystemSuiteJ/RatisMasterStatusSystemSuiteJ,Thanks!

**Comment by leixm** [comment]
cc @SteNicholas @RexXiong

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala [hunk 10]
```diff
@@ -211,4 +211,7 @@ object WorkerSource {
   // active shuffle
   val ACTIVE_SHUFFLE_SIZE = "ActiveShuffleSize"
   val ACTIVE_SHUFFLE_FILE_COUNT = "ActiveShuffleFileCount"
+
+  // decommission
+  val IS_DECOMMISSIONING_WORKER = "isDecommissioningWorker"
```
Comment (SteNicholas) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala:216]:
```suggestion
  val IS_DECOMMISSIONING_WORKER = "IsDecommissioningWorker"
```

File: METRICS.md [hunk 11]
```diff
@@ -147,6 +148,7 @@ Here is an example of Grafana dashboard importing.
 |         PotentialConsumeSpeed          |      worker       |                        This value means speed of potential consumption for congestion control.                         |
 |            UserProduceSpeed            |      worker       |                           This value means speed of user production for congestion control.                            |
 |           WorkerConsumeSpeed           |      worker       |                          This value means speed of worker consumption for congestion control.                          |
+|        isDecommissioningWorker         |      worker       |                              1 means worker decommissioning, 0 means not decommissioning.                              |
```
Comment (SteNicholas) [inline METRICS.md:151]: The descripition of this metric is confusing that 0 not only means not decommissioning, but also mean decommissioned. Right?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala (hunk 10 above)
Comment (leixm) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala:216]: Seems no changed.

File: METRICS.md (hunk 11 above)
Comment (leixm) [reply METRICS.md:151]: If decommissioned the worker will exit, seems  we cannot judge the status of a worker that has exited, but we can judge it through the Master's WorkerCount metrics.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala (hunk 10 above)
Comment (SteNicholas) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala:216]: Are you sure?

Comment (leixm) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala:216]: Sorry I didn't see the case.

Comment (leixm) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala]: Done.

File: METRICS.md (hunk 11 above)
Comment (SteNicholas) [reply METRICS.md]: If this metric only represent that the worker is decommissioning, IMO, we could remove this metric. We could judge this via `DecommissionWorkers` of Master.

Comment (leixm) [reply METRICS.md]: It's ok for me.

Comment (leixm) [reply METRICS.md]: In a production environment, due to certain hardware or environmental reasons, our script will automatically decommission the node. We also need to alert based on isDecommissioning metrics, but we don’t want the alarm to go through graceful shutdown when upgrading the cluster.

Comment (leixm) [reply METRICS.md]: The starting point for mentioning this PR is that we want to distinguish between decommission and graceful shutdown.

Comment (SteNicholas) [reply METRICS.md]: @leixm, where does the worker metric distingish between decommission and graceful shutdown? There is no shutdown metrics in worker.

Comment (leixm) [reply METRICS.md]: The time-consuming process of decommission is longer, I think we will pay more attention to its progress.

Comment (leixm) [reply METRICS.md]: ShutdownWorkerCount metrics includes decommission and graceful shutdown workers.

**Review by waitinfuture — COMMENTED** [review]
LGTM, thanks! BTW, after this PR, `ReportWorkerUnavailable` is only used for `GracefulShutdown`, it's better to change it to `ReportWorkerGracefulShudown`. We can do it in the future :)

**Pull request closed without merging**